src/main/java/uk/ac/ic/wlgitbridge/bridge/lock/PostgresLock.java

```diff
@@ -0,0 +1,132 @@
+package uk.ac.ic.wlgitbridge.bridge.lock;
+
+import uk.ac.ic.wlgitbridge.util.Log;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.Statement;
+
+/**
+ * This ProjectLock class uses a database row lock in Postgres
+ * as its implementation.
+ */
+public class PostgresLock implements ProjectLock {
+
+    private String projectName;
+    private Connection connection;
+    private boolean isLocking;
+    private boolean shouldCommit = false;
+    private OperationTracker registry;
+
+    public PostgresLock(String projectName, OperationTracker operationTracker) {
+      this.projectName = projectName;
+      this.registry = operationTracker;
+      this.isLocking = false;
+    }
+
+    public void setConnection(Connection connection) {
+        this.connection = connection;
+    }
+
+    @Override
+    public void lock() {
+        Log.info("[{}] Trying to get lock", this.projectName);
+        this.registry.lock();
+        try {
+            /**
+             * Turn off auto-commit on this connection, so the next
+             * query will begin a transaction
+             */
+            connection.setAutoCommit(false);
+            /**
+             * "touch" the project row, to ensure there is something to
+             * take a lock on. Does nothing if the row already exists.
+             *
+             * Setting version_id = 0 is important, as it signals that the
+             * project is not "really" present in the system.
+             * @see PostgresDBStore.getProjectState()
+             */
+            try (
+                    PreparedStatement statement = this.connection.prepareStatement(
+                            "INSERT into projects (name, version_id, last_accessed) " +
+                                    "values (?, 0, NOW()) ON CONFLICT DO NOTHING;"
+                    )
+            ) {
+                statement.setString(1, this.projectName);
+                statement.execute();
+            }
+            try (Statement statement = this.connection.createStatement()) {
+                statement.execute("set lock_timeout to 59999;");
```

Comment:
What happens if we do overrun this lock? I some long pushes/pulls do end up taking longer than 1 minute, so we may want to increase this.